packages/yarnpkg-core/sources/Project.ts

```diff
@@ -1351,7 +1351,7 @@ export class Project {
 
           const workspace = this.tryWorkspaceByLocator(dependencyPkg);
           if (workspace) {
-            if (buildablePackages.has(workspace.anchoredLocator.locatorHash))
+            if (resolution !== workspace.anchoredLocator.locatorHash && buildablePackages.has(workspace.anchoredLocator.locatorHash))
```

Comment:
Perhaps I'm too tired and reading this incorrectly, but isn't this checking whether the transitive workspace dependency is different from the devirtualized workspace?

Shouldn't it be?
```suggestion
            if (resolution !== locator.locatorHash && buildablePackages.has(workspace.anchoredLocator.locatorHash))
```

Reply:
You're reading it correctly, that was a silly mistake on my part however your suggested change isn't correct either, the check was supposed to check that the devirtualized workspace wasn't the original `locator`. I changed it to `workspace.anchoredLocator.locatorHash !== locator.locatorHash` and added a new test.